lambdas/es/indexer/index.py

```diff
@@ -94,189 +240,278 @@ def extract_text(notebook_str):
-            use_ssl=True,
-            verify_certs=True,
-            connection_class=RequestsHttpConnection
+def get_plain_text(context, bucket, key, size, *, etag, version_id):
+    """get plain text object contents"""
+    text = ""
+    try:
+        obj = retry_s3(
+            "get",
+            context,
+            bucket,
+            key,
+            size,
+            etag=etag,
+            version_id=version_id
         )
+        text = obj["Body"].read().decode("utf-8")
+    except UnicodeDecodeError as ex:
+        print(f"Unicode decode error in {key}", ex)
+
+    return text
 
-        res = es.index(
-            index=ES_INDEX,
-            doc_type='_doc',
-            body=data,
-            refresh='wait_for'
+def get_time_remaining(context):
+    """returns time remaining in seconds before lambda context is shut down"""
+    time_remaining = floor(context.get_remaining_time_in_millis()/1000)
+    if time_remaining < 30:
+        print(
+            f"Warning: Lambda function has less than {time_remaining} seconds."
+            " Consider reducing bulk batch size."
         )
-        print(res)
-    except RequestError as e:
-        if e.error == 'mapper_parsing_exception':
-            # retry with just plaintext stuff
-            print('Mapping exception. Retrying without user_meta and system_meta')
-            data['user_meta'] = {}
-            data['system_meta'] = {}
-            try:
-                res = es.index(index=ES_INDEX, doc_type='_doc', body=data)
-                print(res)
-            except Exception as e:
-                print('Failover failed. data: ' + json.dumps(data))
-                print(e)
-                import traceback
-                traceback.print_tb(e.__traceback__)
-        else:
-            print("Exception encountered when POSTing to ES")
-            print(e)
-            import traceback
-            traceback.print_tb(e.__traceback__)
 
-    except Exception as e:
-        print("Exception encountered when POSTing to ES")
-        print(e)
-        import traceback
-        traceback.print_tb(e.__traceback__)
+    return time_remaining
+
+def transform_meta(meta):
+    """ Reshapes metadata for indexing in ES """
+    helium = meta.get("helium")
+    user_meta = {}
+    comment = ""
+    target = ""
+
+    if helium:
+        user_meta = helium.pop("user_meta", {})
+        comment = helium.pop("comment", "") or ""
+        target = helium.pop("target", "") or ""
+
+    meta_text_parts = [comment, target]
+
+    if helium:
+        meta_text_parts.append(json.dumps(helium))
+    if user_meta:
+        meta_text_parts.append(json.dumps(user_meta))
 
-def handler(event, _):
-    """fetch the S3 object from event, extract relevant data and metadata,
-    dispatch post_to_es
+    return {
+        "system_meta": helium,
+        "user_meta": user_meta,
+        "comment": comment,
+        "target": target,
+        "meta_text": " ".join(meta_text_parts)
```

Comment:
I never understood this logic... How is this better than just `json.dumps(helium)`? Isn't `meta_text_parts` the same thing, just slightly flatter?

Reply:
i think it has do with the fact that the JSON characters would inhibit plain text indexing of comment and target